File: src/game/buildings/growth.rs
use bevy::ecs::system::SystemParam;
use bevy::prelude::*;
use rand::{Rng, SeedableRng, rngs::StdRng};
use std::collections::HashSet;

use crate::game::demand::RciDemand;
use crate::game::land_value::LandValueIndex;
use crate::game::map::{BuildingKind, DirtyTiles, MapConfig, MapEditVersion, MapGrid, TilePos};
use crate::game::notifications::{NotificationKind, Notifications};
use crate::game::sim::City;
use crate::game::sim_events::HourAdvanced;
use bevy::ecs::message::MessageReader;

use super::components::*;
use super::spawn::spawn_building_entity;
use super::zone_depth::{MAX_ZONE_DEPTH, is_footprint_within_zone_depth};

#[derive(SystemParam)]
pub struct GrowBuildingsParams<'w, 's> {
    hour_events: MessageReader<'w, 's, HourAdvanced>,
    cfg: Res<'w, MapConfig>,
    grid: ResMut<'w, MapGrid>,
    demand: Res<'w, RciDemand>,
    land_value: Option<Res<'w, LandValueIndex>>,
    city: ResMut<'w, City>,
    rng: ResMut<'w, BuildingGrowthRng>,
    dirty: ResMut<'w, DirtyTiles>,
    /// Tracks map edits when buildings grow.
    map_edit_version: ResMut<'w, MapEditVersion>,
    commands: Commands<'w, 's>,
    q_buildings: Query<'w, 's, &'static Building>,
    notifications: Option<ResMut<'w, Notifications>>,
}

/// GDD: Building growth happens every 1 game hour (via HourAdvanced event)
pub fn grow_buildings(mut p: GrowBuildingsParams) {
    // Trigger on every hour advanced (GDD: each 1 game hour)
    if p.hour_events.read().count() == 0 {
        return;
    }

    // Limit growth per tick (MVP).
    let max_spawns = 6usize;
    let mut spawned = 0usize;

    // Count currently under-construction buildings (GDD 10.3.3.1)
    let under_construction_count = p
        .q_buildings
        .iter()
        .filter(|b| {
            matches!(
                b.phase,
                super::components::BuildingPhase::UnderConstruction { .. }
            )
        })
        .count();

    // Limit parallel constructions (GDD 10.3.3.1: city-wide construction capacity)
    const MAX_PARALLEL_CONSTRUCTIONS: usize = 15;
    if under_construction_count >= MAX_PARALLEL_CONSTRUCTIONS {
        return; // Don't spawn new buildings if construction capacity is full
    }

    // Build a quick occupancy set to avoid double-spawning if something went out of sync.
    // Track all tiles occupied by building footprints
    let mut occupied = HashSet::<TilePos>::new();
    for b in &p.q_buildings {
        for tile in b.footprint_tiles() {
            occupied.insert(tile);
        }
    }

    let len = p.grid.len();
    if len == 0 {
        return;
    }

    // Random attempts instead of full scan (cheap + scalable).
    for _ in 0..128 {
        if spawned >= max_spawns {
            break;
        }

        let idx = p.rng.rng.random_range(0..len);
        let x = (idx % (p.grid.width as usize)) as i32;
        let y = (idx / (p.grid.width as usize)) as i32;
        let seed_pos = TilePos { x, y };

        // Check if seed tile is valid for building
        let Some(cell) = p.grid.get(seed_pos) else {
            continue;
        };
        if cell.water || cell.road.is_some() || cell.building.is_some() {
            continue;
        }

        let Some(kind) = BuildingKind::from_zone(cell.zone) else {
            continue;
        };
        if !demand_allows_growth(&p.demand, kind) {
            continue;
        }

        // Try to find a valid footprint starting from this seed position
        let Some(footprint) = find_best_footprint(
            seed_pos,
            kind,
            &p.grid,
            &occupied,
            p.land_value.as_deref(),
            &p.q_buildings,
        ) else {
            continue;
        };

        // Mark all footprint tiles in sim state (source of truth)
        let mut footprint_tiles = Vec::new();
        for tile in footprint.tiles.iter() {
            if let Some(mut cell) = p.grid.get(*tile) {
                cell.building = Some(kind);
                p.grid.set(*tile, cell);
                if let Some(idx) = p.grid.idx(*tile) {
                    p.dirty.mark(idx);
                }
                footprint_tiles.push(*tile);
            }
        }

        // Spawn render entity with footprint
        spawn_building_entity(
            &mut p.commands,
            &p.cfg,
            footprint.anchor,
            footprint.width,
            footprint.length,
            kind,
            &p.city,
        );
        p.map_edit_version.bump();

        // Mark all footprint tiles as occupied
        for tile in &footprint_tiles {
            occupied.insert(*tile);
        }
        spawned += 1;

        // Emit notification
        if let Some(ref mut notif) = p.notifications {
            let kind_name = match kind {
                BuildingKind::Residential => "Residential",
                BuildingKind::Commercial => "Commercial",
                BuildingKind::Industrial => "Industrial",
                _ => "Building",
            };
            notif.add(
                format!("New {} building constructed", kind_name),
                NotificationKind::Info,
                3.0,
            );
        }

        // Population is now calculated from occupancy, not added here
        // Occupancy will be updated by the occupancy system when the building becomes operational
    }
}

fn demand_allows_growth(demand: &RciDemand, kind: BuildingKind) -> bool {
    match kind {
        BuildingKind::Residential => demand.residential > 0.0,
        BuildingKind::Commercial => demand.commercial > 0.0,
        BuildingKind::Industrial => demand.industrial > 0.0,
        _ => true,
    }
}

/// Represents a valid building footprint
struct Footprint {
    anchor: TilePos,
    width: u8,
    length: u8,
    tiles: Vec<TilePos>,
}

/// Find the best valid footprint starting from a seed position.
/// Algorithm from GDD 10.1.3: priority is area → length → width.
/// Tries footprints from 6x6 down to 3x3, checking all orientations.
fn find_best_footprint(
    seed_pos: TilePos,
    kind: BuildingKind,
    grid: &MapGrid,
    occupied: &HashSet<TilePos>,
    land_value: Option<&LandValueIndex>,
    existing_buildings: &Query<&Building>,
) -> Option<Footprint> {
    // Generate all possible footprints sorted by priority: area → length → width
    // GDD 10.1.3: priority is area (desc), then length (desc), then width (desc)
    // This means: 6x6, then 6x5, 5x6, then 6x4, 4x6, 5x5, then 6x3, 3x6, 5x4, 4x5, etc.
    let mut candidates = Vec::new();
    // Generate all valid (width, length) pairs where both are 3-6
    for w in 3..=6 {
        for l in 3..=6 {
            candidates.push((w as u8, l as u8));
        }
    }

    // Sort by priority: area (desc), then length (desc), then width (desc)
    candidates.sort_by(|a, b| {
        let area_a = (a.0 as u32) * (a.1 as u32);
        let area_b = (b.0 as u32) * (b.1 as u32);
        area_b
            .cmp(&area_a)
            .then_with(|| b.1.cmp(&a.1)) // length desc
            .then_with(|| b.0.cmp(&a.0)) // width desc
    });

    // Try each candidate footprint, checking if it can be placed starting from seed_pos
    for (width, length) in candidates {
        // Try placing the footprint with seed_pos as anchor
        if let Some(footprint) = try_footprint_at(
            seed_pos,
            width,
            length,
            kind,
            grid,
            occupied,
            land_value,
            existing_buildings,
        ) {
            return Some(footprint);
        }

        // Also try placing with seed_pos at different corners of the footprint
        // This allows finding footprints that contain the seed but don't start at it
        for anchor_offset in [
            (0, 0),
            (-(width as i32 - 1), 0),
            (0, -(length as i32 - 1)),
            (-(width as i32 - 1), -(length as i32 - 1)),
        ] {
            let anchor = TilePos {
                x: seed_pos.x + anchor_offset.0,
                y: seed_pos.y + anchor_offset.1,
            };
            if let Some(footprint) = try_footprint_at(
                anchor,
                width,
                length,
                kind,
                grid,
                occupied,
                land_value,
                existing_buildings,
            ) {
                return Some(footprint);
            }
        }
    }

    None
}

/// Try to place a footprint at a specific anchor position.
/// Returns Some(Footprint) if valid, None otherwise.
/// GDD 10.1.3: RCI buildings cannot be built "behind" other buildings (blocking rule).
fn try_footprint_at(
    anchor: TilePos,
    width: u8,
    length: u8,
    kind: BuildingKind,
    grid: &MapGrid,
    occupied: &HashSet<TilePos>,
    land_value: Option<&LandValueIndex>,
    existing_buildings: &Query<&Building>,
) -> Option<Footprint> {
    let mut tiles = Vec::new();
    let required_zone = kind.as_zone();

    // Check all tiles in the footprint
    for dx in 0..(width as i32) {
        for dy in 0..(length as i32) {
            let tile = TilePos {
                x: anchor.x + dx,
                y: anchor.y + dy,
            };

            // Check if tile is valid
            let Some(cell) = grid.get(tile) else {
                return None;
            };

            // Check basic constraints
            if cell.water || cell.road.is_some() || cell.building.is_some() {
                return None;
            }

            // Check zone matches
            if cell.zone != required_zone {
                return None;
            }

            // Check not occupied by another building footprint
            if occupied.contains(&tile) {
                return None;
            }

            tiles.push(tile);
        }
    }

    // Check zone depth (GDD 10.2.2): all tiles must be within 3 tiles of a road
    if !is_footprint_within_zone_depth(anchor, width, length, grid, MAX_ZONE_DEPTH) {
        return None;
    }

    // Check road access: at least one tile must have adjacent road (GDD: all buildings must be adjacent to road)
    let mut has_road_access = false;
    let mut road_adjacent_tiles = Vec::new();
    for tile in &tiles {
        if has_adjacent_road(grid, *tile) {
            has_road_access = true;
            road_adjacent_tiles.push(*tile);
        }
    }
    if !has_road_access {
        return None;
    }

    // GDD 10.1.3: Check blocking rule - RCI buildings cannot be built "behind" other RCI buildings
    // (Service buildings are exempt from this rule)
    // Simplified approach: if a tile in the new footprint is further from the nearest road than
    // any existing RCI building on the same "line" from that road, it's blocked
    if matches!(
        kind,
        BuildingKind::Residential | BuildingKind::Commercial | BuildingKind::Industrial
    ) {
        // For each road-adjacent tile in the new footprint, check if there's a blocking building
        for road_tile in &road_adjacent_tiles {
            // Find all existing RCI buildings that are adjacent to this same road tile
            for building in existing_buildings.iter() {
                // Only check RCI buildings (not service buildings)
                if !matches!(
                    building.kind,
                    BuildingKind::Residential | BuildingKind::Commercial | BuildingKind::Industrial
                ) {
                    continue;
                }

                let building_tiles = building.footprint_tiles();
                let building_has_this_road = building_tiles.iter().any(|bt| {
                    // Check if building tile is adjacent to the same road tile
                    (bt.x - road_tile.x).abs() + (bt.y - road_tile.y).abs() == 1
                });

                if building_has_this_road {
                    // Check if any tile in the new footprint is "behind" this building
                    // (further from the road in the same direction)
                    for new_tile in &tiles {
                        // Skip if new tile is also adjacent to road (it's at the front, not behind)
                        if has_adjacent_road(grid, *new_tile) {
                            continue;
                        }

                        // Check if new tile is in the same "direction" from road as the building
                        // Simple heuristic: if building extends in a direction, new tile shouldn't be further in that direction
                        let building_max_dist = building_tiles
                            .iter()
                            .map(|bt| {
                                let dx = bt.x - road_tile.x;
                                let dy = bt.y - road_tile.y;
                                dx.abs() + dy.abs()
                            })
                            .max()
                            .unwrap_or(0);

                        let new_tile_dist =
                            (new_tile.x - road_tile.x).abs() + (new_tile.y - road_tile.y).abs();

                        // If new tile is further from road than building extends, it might be blocked
                        // But we need to check if they're on the same "line"
                        // Simplified: if building is between road and new tile, block it
                        if new_tile_dist > building_max_dist {
                            // Check if new tile is in the same quadrant/direction from road
                            let building_dir_x = if building.anchor_pos.x > road_tile.x {
                                1
                            } else if building.anchor_pos.x < road_tile.x {
                                -1
                            } else {
                                0
                            };
                            let building_dir_y = if building.anchor_pos.y > road_tile.y {
                                1
                            } else if building.anchor_pos.y < road_tile.y {
                                -1
                            } else {
                                0
                            };
                            let new_dir_x = if new_tile.x > road_tile.x {
                                1
                            } else if new_tile.x < road_tile.x {
                                -1
                            } else {
                                0
                            };
                            let new_dir_y = if new_tile.y > road_tile.y {
                                1
                            } else if new_tile.y < road_tile.y {
                                -1
                            } else {
                                0
                            };

                            // If directions match (same quadrant), block it
                            if building_dir_x == new_dir_x && building_dir_y == new_dir_y {
                                return None;
                            }
                        }
                    }
                }
            }
        }
    }

    // Check land value requirement (use minimum value from all tiles in footprint)
    if let Some(land_val) = land_value {
        let min_value = match kind {
            BuildingKind::Residential => 0.3,
            BuildingKind::Commercial => 0.4,
            BuildingKind::Industrial => 0.0,
            _ => 0.0,
        };
        if min_value > 0.0 {
            let mut all_valid = true;
            for tile in &tiles {
                if let Some(idx) = grid.idx(*tile) {
                    let value = land_val.get(idx);
                    if value < min_value {
                        all_valid = false;
                        break;
                    }
                }
            }
            if !all_valid {
                return None;
            }
        }
    }

    Some(Footprint {
        anchor,
        width,
        length,
        tiles,
    })
}

fn has_adjacent_road(grid: &MapGrid, pos: TilePos) -> bool {
    let neighbors = [
        TilePos {
            x: pos.x - 1,
            y: pos.y,
        },
        TilePos {
            x: pos.x + 1,
            y: pos.y,
        },
        TilePos {
            x: pos.x,
            y: pos.y - 1,
        },
        TilePos {
            x: pos.x,
            y: pos.y + 1,
        },
    ];

    for npos in neighbors {
        if let Some(cell) = grid.get(npos)
            && cell.road.is_some()
        {
            return true;
        }
    }
    false
}

/// Find the nearest road tile to a given position using BFS (for blocking rule check)
fn find_nearest_road_tile(grid: &MapGrid, start: TilePos) -> Option<TilePos> {
    use std::collections::VecDeque;
    let mut queue = VecDeque::new();
    let mut visited = HashSet::new();
    queue.push_back(start);
    visited.insert(start);

    while let Some(current) = queue.pop_front() {
        if has_adjacent_road(grid, current) {
            // Return the road tile itself (one of the adjacent tiles)
            for (dx, dy) in [(-1, 0), (1, 0), (0, -1), (0, 1)] {
                let neighbor = TilePos {
                    x: current.x + dx,
                    y: current.y + dy,
                };
                if let Some(cell) = grid.get(neighbor) {
                    if cell.road.is_some() {
                        return Some(neighbor);
                    }
                }
            }
        }

        for (dx, dy) in [(-1, 0), (1, 0), (0, -1), (0, 1)] {
            let neighbor = TilePos {
                x: current.x + dx,
                y: current.y + dy,
            };
            if !visited.contains(&neighbor) && grid.get(neighbor).is_some() {
                visited.insert(neighbor);
                queue.push_back(neighbor);
            }
        }
    }
    None
}

pub fn seed_growth_rng_from_map(
    seed: Res<crate::game::map::MapSeed>,
    mut rng: ResMut<BuildingGrowthRng>,
) {
    rng.rng = StdRng::seed_from_u64(seed.0);
}

pub fn reset_growth_rng_on_new_map(
    mut reader: bevy::ecs::message::MessageReader<crate::game::commands::GameCommand>,
    seed: Res<crate::game::map::MapSeed>,
    mut rng: ResMut<BuildingGrowthRng>,
) {
    for cmd in reader.read() {
        if matches!(cmd, crate::game::commands::GameCommand::GenerateMap { .. }) {
            rng.rng = StdRng::seed_from_u64(seed.0);
        }
    }
}

File: packages/sim/src/buildings/growth.ts
// Port of crates/simcity_sim/src/game/buildings/growth.rs: every game hour zoned land with demand, a road
// and power grows buildings at random seeds, the largest footprint that fits first (GDD 10.1.3).
import { attractivenessToGrow, workplaceClass } from '../cityFields';
import type { BuildingKind, TilePos, ZoneDensity } from '../commands';
import { zoneDemand } from '../demand';
import { wealthFromLandValue } from '../economy/wealth';
import { bumpVersion } from '../map/dirty';
import { tileKey } from '../map/grid';
import { roadCellIsSome } from '../map/roads';
import { MAX_ZONE_DEPTH, isFootprintWithinZoneDepth } from '../map/zonePlacement';
import { rangeU32 } from '../rng';
import type { World } from '../world';
import {
  anyFootprintTile,
  buildingKindFromZone,
  buildingKindZone,
  footprintSides,
  footprintTiles,
  hasAdjacentRoad,
  isZonedKind,
  type Building,
} from './building';
import { spawnBuilding } from './spawn';

/** One feed line for every zone. */
export const CONSTRUCTION_NOTICE = 'New building constructed';
const MAX_SPAWNS_PER_HOUR = 6;
/** GDD 10.3.3.1: the city's construction capacity. */
const MAX_PARALLEL_CONSTRUCTIONS = 15;
const SEED_ATTEMPTS = 128;

interface Footprint {
  readonly anchor: TilePos;
  readonly width: number;
  readonly length: number;
  readonly tiles: readonly TilePos[];
}

const sign = (v: number) => (v > 0 ? 1 : v < 0 ? -1 : 0);

/**
 * GDD 10.1.3 blocking rule: a zoned building does not grow behind another zoned building along the same road
 * tile, further from the road in the same direction.
 */
function blockedBehind(w: World, tiles: readonly TilePos[], existing: readonly Building[]): boolean {
  const manhattan = (a: TilePos, b: TilePos) => Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
  for (const roadTile of tiles.filter((tile) => hasAdjacentRoad(w.grid, tile))) {
    for (const b of existing) {
      if (!isZonedKind(b.kind)) continue;
      if (!anyFootprintTile(b.anchor, b.width, b.length, (bt) => manhattan(bt, roadTile) === 1)) continue;
      const reach = Math.max(...footprintTiles(b).map((bt) => manhattan(bt, roadTile)));
      for (const tile of tiles) {
        if (hasAdjacentRoad(w.grid, tile) || manhattan(tile, roadTile) <= reach) continue;
        if (sign(b.anchor.x - roadTile.x) === sign(tile.x - roadTile.x) && sign(b.anchor.y - roadTile.y) === sign(tile.y - roadTile.y)) return true;
      }
    }
  }
  return false;
}

function tryFootprintAt(
  w: World,
  anchor: TilePos,
  width: number,
  length: number,
  kind: BuildingKind,
  density: ZoneDensity,
  occupied: ReadonlySet<number>,
  existing: readonly Building[],
): Footprint | undefined {
  const grid = w.grid;
  const zone = buildingKindZone(kind);
  const tiles: TilePos[] = [];
  for (let dx = 0; dx < width; dx++) {
    for (let dy = 0; dy < length; dy++) {
      const tile = { x: anchor.x + dx, y: anchor.y + dy };
      const cell = grid.get(tile);
      if (cell === undefined || cell.water || roadCellIsSome(cell.road) || cell.building !== null) return undefined;
      if (cell.zone !== zone || cell.density !== density || occupied.has(tileKey(tile))) return undefined;
      tiles.push(tile);
    }
  }
  // GDD 10.2.2: every tile within zone depth of a road, and at least one beside a road.
  if (!isFootprintWithinZoneDepth(anchor, width, length, grid, MAX_ZONE_DEPTH)) return undefined;
  if (!tiles.some((tile) => hasAdjacentRoad(grid, tile))) return undefined;
  // Nothing grows where the road carries no power.
  if (!w.utilityNetwork.footprintHas(grid, anchor, width, length, 'Power')) return undefined;
  if (blockedBehind(w, tiles, existing)) return undefined;

  // Once the city fields are measured a zone grows where it is attractive enough on every tile; before
  // that, bare land value decides.
  const fields = w.cityFields;
  if (fields.covers(grid.len())) {
    const floor = attractivenessToGrow(kind);
    if (floor !== undefined && tiles.some((tile) => fields.get('Attractiveness', grid.idx(tile)!) < floor)) return undefined;
  } else {
    const min = kind === 'Residential' ? Math.fround(0.3) : kind === 'Commercial' ? Math.fround(0.4) : 0;
    if (min > 0 && tiles.some((tile) => w.landValue.get(grid.idx(tile)!) < min)) return undefined;
  }
  return { anchor, width, length, tiles };
}

/** Footprints by area, then length, then width, each tried with the seed at its four corners. */
function findBestFootprint(
  w: World,
  seed: TilePos,
  kind: BuildingKind,
  density: ZoneDensity,
  occupied: ReadonlySet<number>,
  existing: readonly Building[],
): Footprint | undefined {
  const [shortest, longest] = footprintSides(density);
  const sizes: Array<readonly [number, number]> = [];
  for (let width = shortest; width <= longest; width++) {
    for (let length = shortest; length <= longest; length++) sizes.push([width, length]);
  }
  sizes.sort((a, b) => b[0] * b[1] - a[0] * a[1] || b[1] - a[1] || b[0] - a[0]);
  for (const [width, length] of sizes) {
    for (const [ox, oy] of [
      [0, 0],
      [0, 0],
      [-(width - 1), 0],
      [0, -(length - 1)],
      [-(width - 1), -(length - 1)],
    ] as const) {
      const found = tryFootprintAt(w, { x: seed.x + ox, y: seed.y + oy }, width, length, kind, density, occupied, existing);
      if (found !== undefined) return found;
    }
  }
  return undefined;
}

/** `grow_buildings` (SimStep::Buildings), on every advanced hour. */
export function growBuildings(w: World): void {
  if (w.events.hourAdvanced.length === 0) return;
  // New buildings join the rules from the next hour on, as spawned entities did in Rust.
  const existing = [...w.buildings.all()];
  if (existing.filter((b) => b.phase.kind === 'UnderConstruction').length >= MAX_PARALLEL_CONSTRUCTIONS) return;
  const grid = w.grid;
  const len = grid.len();
  if (len === 0) return;
  const occupied = new Set<number>();
  for (const b of existing) for (const tile of footprintTiles(b)) occupied.add(tileKey(tile));

  let spawned = 0;
  for (let attempt = 0; attempt < SEED_ATTEMPTS && spawned < MAX_SPAWNS_PER_HOUR; attempt++) {
    const idx = rangeU32(w.growthRng, 0, len);
    const seed = { x: idx % grid.width, y: Math.trunc(idx / grid.width) };
    const cell = grid.cellAt(idx);
    if (cell.water || roadCellIsSome(cell.road) || cell.building !== null) continue;
    const kind = buildingKindFromZone(cell.zone);
    if (kind === undefined || !(zoneDemand(w.rciDemand, kind) > 0)) continue;
    const footprint = findBestFootprint(w, seed, kind, cell.density, occupied, existing);
    if (footprint === undefined) continue;

    for (const tile of footprint.tiles) {
      grid.set(tile, { ...grid.get(tile)!, building: kind });
      w.dirty.mark(grid.idx(tile)!);
      occupied.add(tileKey(tile));
    }
    // The class comes from the land; a workplace takes the high class only where the people around are educated for it.
    const land = w.landValue.values.length === len ? wealthFromLandValue(w.landValue.get(grid.idx(footprint.anchor)!)) : 'Middle';
    const wealth =
      kind === 'Commercial' || kind === 'Industrial'
        ? workplaceClass(land, w.cityFields.footprintMean('Education', grid, footprint.anchor, footprint.width, footprint.length))
        : land;
    spawnBuilding(w, footprint.anchor, footprint.width, footprint.length, kind, false, { density: cell.density, class: wealth });
    spawned += 1;
    w.notifications.addAt(CONSTRUCTION_NOTICE, 'Info', 3, footprint.anchor);
  }
  // The renderer and the utility network read the building layer by the edit version.
  if (spawned > 0) w.mapEditVersion = bumpVersion(w.mapEditVersion);
}
